### PR TITLE
Add deq module

### LINKED ISSUE #25: Implement Deq module
Implement persistent real-time double-ended queues, using the techniques described in __Purely Functional Data Structures__, by Chris Okasaki.  The `Deq` API should have substantial overlap with the `List` API, with at least the following additions:

- `val push_back: 'a t -> 'a -> 'a t`: Push element onto the back and return the updated dequeue.
- `val back: 'a t -> 'a`: Return the back element.
- `val pop_back: 'a t -> 'a * 'a t`: Pop the back element and return it, along with the updated dequeue.

### PR DESCRIPTION
**Pull request by cevans87**
Close #25

The implementation for this module was adapted from the real-time
double-ended queue in the book "Purely Functional Data Structures" by
Chris Okasaki.

### CODE REVIEW

File: bootstrap/src/deq.ml [hunk 1]
```diff
@@ -0,0 +1,320 @@
+open Rudiments
+
+type 'a t =
+  usize * 'a Stream.t * 'a Stream.t * usize * 'a Stream.t * 'a Stream.t
+
+let c = 2
```
Comment (jasone) [inline bootstrap/src/deq.ml]: A brief comment here would be useful to make it clear this isn't just a sloppy magic number.  Something like:

```ocaml
(* Increment size for rebuilding; must be either 2 or 3. *)
```

File: bootstrap/src/deq.ml [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+    | lazy Stream.Nil -> ()
+    | lazy (Cons(elm, lazy Stream.Nil)) -> fprintf ppf "%a" pp_elm elm;
+    | lazy (Cons(elm, s')) -> begin
+      fprintf ppf "%a,@ " pp_elm elm;
+      fn s'
+    end
+  end in
+  fprintf ppf "(";
+  fn f;
+  fprintf ppf "@ |@ ";
```
Comment (jasone) [inline bootstrap/src/deq.ml]: The `( ... | ... )` format surprised me out at first, but I really like it.

File: bootstrap/src/deq.mli [hunk 3]
```diff
@@ -0,0 +1,59 @@
+(** Lazy-evaluated immutable double-ended queue.
+
+     The deq is an immutable double-ended queue container. All operations on deq
+     are amortized via lazy execution and are O(1).
```
Comment (jasone) [inline bootstrap/src/deq.mli]: It looks like this might be over-indented.

**Comment by cevans87** [comment]
I moved "Close #25" to the bottom. It shows in the new commit message but not at the top in this discussion.

**Pull request merged**